### PR TITLE
Fix copying shared libraries after build

### PR DESCRIPTION
**Pull request by ryanking13**
### Description

This fixes two minor bugs when building packages:

1. Shared libraries are not copied into dist directory when it is already built.
2. When build fails, the shared library zip file remains in the package directory.

After this PR, the zip file will be created in `packages/<shared-lib>/dist` similar to how we place wheel files.

This PR fixes the test failure in #3192

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry

### CODE REVIEW

File: pyodide-build/pyodide_build/buildpkg.py [hunk 1]
```diff
@@ -733,8 +733,12 @@ def build_package(
     name = pkg.package.name
     version = pkg.package.version
     build_dir = pkg_root / "build"
+    dist_dir = pkg_root / "dist"
     src_dir_name: str = f"{name}-{version}"
     srcpath = build_dir / src_dir_name
+    src_dist_dir = (
+        srcpath / "dist"
+    )  # build output (.whl, .so) will be go here, then copied to dist_dir
```
Comment (hoodmane) [inline pyodide-build/pyodide_build/buildpkg.py:741]:
```suggestion
    src_dist_dir = srcpath / "dist"
    # Python produces output .whl or .so files in src_dist_dir.
    # We copy them to dist_dir later
```

File: pyodide-build/pyodide_build/buildpkg.py [hunk 2]
```diff
@@ -784,27 +788,29 @@ def build_package(
...
-            )
-        if not sharedlibrary:
+            # Nothing needs to be done for a static library
+            pass
+        elif sharedlibrary:
+            # If shared library, we copy .so files to dist_dir
+            # and create a zip archive of the .so files
+            shutil.rmtree(dist_dir, ignore_errors=True)
+            dist_dir.mkdir(parents=True)
+            shutil.make_archive(str(dist_dir / src_dir_name), "zip", srcpath / "dist")
```
Comment (hoodmane) [inline pyodide-build/pyodide_build/buildpkg.py:798]: It's a bit hard to keep track of all these similar combinations of things...
```suggestion
            shutil.make_archive(str(dist_dir / src_dir_name), "zip", src_dist_dir)
```

File: pyodide-build/pyodide_build/tests/test_buildall.py [hunk 3]
```diff
@@ -52,7 +52,9 @@ def from_yaml(cls, path):
 
 
 def test_generate_repodata(tmp_path):
-    pkg_map = buildall.generate_dependency_graph(PACKAGES_DIR, {"pkg_1", "pkg_2"})
+    pkg_map = buildall.generate_dependency_graph(
+        PACKAGES_DIR, {"pkg_1", "pkg_2", "libtest"}
```
Comment (hoodmane) [inline pyodide-build/pyodide_build/tests/test_buildall.py:56]: I assume adding libtest here demonstrates the bug you fixed? Why?

File: pyodide-build/pyodide_build/buildall.py [hunk 4]
```diff
@@ -559,10 +556,10 @@ def copy_packages_to_dist_dir(
     packages: Iterable[BasePackage], output_dir: Path
 ) -> None:
     for pkg in packages:
-        try:
-            shutil.copy(pkg.wheel_path(), output_dir)
-        except RuntimeError:
-            pass
+        if pkg.library:
+            continue
+
```
Comment (hoodmane) [inline pyodide-build/pyodide_build/buildall.py:561]: Much better. This `except RuntimeError: pass` is pretty horrendous (so I probably wrote it...)

**Review by hoodmane — APPROVED** [review]
Thanks @ryanking13!

File: pyodide-build/pyodide_build/buildpkg.py (hunk 2 above)
Comment (ryanking13) [reply pyodide-build/pyodide_build/buildpkg.py:798]: Oh, thanks. I forgot this one.

File: pyodide-build/pyodide_build/tests/test_buildall.py (hunk 3 above)
Comment (ryanking13) [reply pyodide-build/pyodide_build/tests/test_buildall.py:56]: No, this one is not related to the bug. I added this to check that static libraries are not included in the `repodata.json`. which was not covered by previous tests.

**Pull request merged**